### PR TITLE
added logs in realtime + save to DB

### PR DESCRIPTION
**Pull request by shay23b**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -15,34 +15,38 @@ package main
 
 import (
 	"memphis-control-plane/analytics"
+	"memphis-control-plane/background_tasks"
```
Comment (idanasulinmemphis) [inline main.go:18]: move this import to be after the DB import

File: main.go [hunk 2]
```diff
@@ -15,34 +15,38 @@ package main
...
 	err := analytics.InitializeAnalytics()
-	handleError("Failed to initialize analytics:", err)
+	handleError("Failed to initialize analytics: ", err)
 
 	err = handlers.CreateRootUserOnFirstSystemLoad()
-	handleError("Failed to create root user:", err)
+	handleError("Failed to create root user: ", err)
+
+	err = logger.InitializeLogger()
+	handleError("Failed initializing logger", err)
```
Comment (idanasulinmemphis) [inline main.go:37]: add : + space

File: middlewares/auth.go [hunk 3]
```diff
@@ -14,13 +14,14 @@
...
 
 import (
-	"context"
-	"errors"
-	"fmt"
 	"memphis-control-plane/config"
 	"memphis-control-plane/db"
 	"memphis-control-plane/logger"
 	"memphis-control-plane/models"
+
```
Comment (idanasulinmemphis) [inline middlewares/auth.go:21]: change it back

File: handlers/consumers.go [hunk 4]
```diff
@@ -14,12 +14,13 @@
 package handlers
 
 import (
-	"context"
-	"errors"
 	"memphis-control-plane/broker"
 	"memphis-control-plane/logger"
 	"memphis-control-plane/models"
 	"memphis-control-plane/utils"
+
```
Comment (idanasulinmemphis) [inline handlers/consumers.go:21]: change it back

File: broker/broker.go [hunk 5]
```diff
@@ -14,20 +14,26 @@
 package broker
 
 import (
-	"errors"
 	"memphis-control-plane/config"
-	"memphis-control-plane/logger"
+	"memphis-control-plane/db"
```
Comment (idanasulinmemphis) [inline broker/broker.go:18]: db logic should be inside the consumeLogs section

File: broker/broker.go [hunk 6]
```diff
@@ -14,20 +14,26 @@
...
 	"memphis-control-plane/models"
+
+	"errors"
+	"log"
 	"strings"
 	"time"
 
 	"github.com/nats-io/nats.go"
 	"github.com/nats-io/nkeys"
+	"go.mongodb.org/mongo-driver/bson/primitive"
```
Comment (idanasulinmemphis) [inline broker/broker.go:28]: not needed here

File: broker/broker.go [hunk 7]
```diff
@@ -278,6 +284,73 @@ func ValidateUserCreds(token string) error {
 	return nil
 }
 
+func CreateInternalStream(name string, subjects []string) error {
```
Comment (idanasulinmemphis) [inline broker/broker.go:287]: get only the name of the stream and use it as the only subject

File: broker/broker.go [hunk 8]
```diff
@@ -278,6 +284,73 @@ func ValidateUserCreds(token string) error {
...
+		Retention: nats.WorkQueuePolicy,
+		MaxConsumers: -1,
+		Storage: nats.FileStorage,
+		Replicas: 1,
+		NoAck: false,
+		Duplicates: dedupWindow,
+	}, nats.MaxWait(10*time.Second))
+	if err != nil {
+		if !strings.Contains(err.Error(), "invalid bucket name"){
+			return getErrorWithoutNats(err)
```
Comment (idanasulinmemphis) [inline broker/broker.go:301]: this if statement should check whether a Stream with that name is already exist and if it does => ignore this error, not return an error

File: broker/broker.go [hunk 9]
```diff
@@ -278,6 +284,73 @@ func ValidateUserCreds(token string) error {
...
+		Name: "$memphis_" + name,
+		Subjects: subjects,
+		Retention: nats.WorkQueuePolicy,
+		MaxConsumers: -1,
+		Storage: nats.FileStorage,
+		Replicas: 1,
+		NoAck: false,
+		Duplicates: dedupWindow,
+	}, nats.MaxWait(10*time.Second))
+	if err != nil {
```
Comment (idanasulinmemphis) [inline broker/broker.go:299]: combine those conditions into 1 condition

File: broker/broker.go [hunk 10]
```diff
@@ -278,6 +284,73 @@ func ValidateUserCreds(token string) error {
...
+		Component: component,
+	}
+	_, err := js.Publish(stream, models.Log.ToBytes(log))
+	if err != nil {
+		return getErrorWithoutNats(err)
+	}
+	return nil
+}
+
+func CreateSubscriber(stream string, durable string) (*nats.Subscription, error) {
```
Comment (idanasulinmemphis) [inline broker/broker.go:346]: change to createpullsubscriber

File: logger/logger.go [hunk 11]
```diff
@@ -14,27 +14,53 @@
...
-	"log"
+	"memphis-control-plane/broker"
 	"memphis-control-plane/config"
+
+	"log"
 )
 
-var logger = log.Default()
 var configuration = config.GetConfig()
+var connectionChannel = make(chan bool)
```
Comment (idanasulinmemphis) [inline logger/logger.go:24]: remove these lines

File: background_tasks/logs_consumer.go [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+	"time"
+
+	"github.com/nats-io/nats.go"
+	"go.mongodb.org/mongo-driver/mongo"
+)
+
+var sysLogsCollection *mongo.Collection = db.GetCollection("system_logs")
+
+func ConsumeLogs(wg *sync.WaitGroup) {
+	sub, err := broker.CreateSubscriber("$memphis_sys_logs", "subLog")
```
Comment (idanasulinmemphis) [inline background_tasks/logs_consumer.go:35]: change subLog to logs_consumers

File: background_tasks/logs_consumer.go [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+	if err != nil {
+		logger.Error("Failed creating log subscriber: " + err.Error())
+	}
+
+	defer wg.Done()
```
Comment (idanasulinmemphis) [inline background_tasks/logs_consumer.go:40]: move to the top of this function

File: background_tasks/logs_consumer.go [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+	for range time.Tick(time.Second * 10) {
+		msgs, err := sub.Fetch(1000, nats.MaxWait(10*time.Second))
+		if err != nil {
+			if !strings.Contains(err.Error(), "timeout"){
```
Comment (idanasulinmemphis) [inline background_tasks/logs_consumer.go:45]: add comment which explains what this line is checking

File: background_tasks/logs_consumer.go [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+		if len(msgs) > 0 {
+			msgLogs := make([]interface{}, len(msgs))
+			logsForSocket := make([]models.Log, len(msgs))
+			var singleLog models.Log
+			for index, msg := range msgs{
+				err := json.Unmarshal(msg.Data, &singleLog)
+				if err != nil {
+					logger.Error("Error converting logs: " + err.Error())
+				}
+				msgLogs[index] = singleLog
```
Comment (idanasulinmemphis) [inline background_tasks/logs_consumer.go:59]: this lines should execute only when no errors occured

File: background_tasks/logs_consumer.go [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+	for range time.Tick(time.Second * 10) {
+		msgs, err := sub.Fetch(1000, nats.MaxWait(10*time.Second))
+		if err != nil {
+			if !strings.Contains(err.Error(), "timeout"){
+				logger.Error("Error fetching logs: " + err.Error())
+			}
+		}
+		
+		if len(msgs) > 0 {
+			msgLogs := make([]interface{}, len(msgs))
```
Comment (idanasulinmemphis) [inline background_tasks/logs_consumer.go:51]: change from msgLogs to logsForDB

File: socketio/socketio.go [hunk 17]
```diff
@@ -14,24 +14,33 @@
 package socketio
 
 import (
-	"errors"
+	"log"
```
Comment (idanasulinmemphis) [inline socketio/socketio.go:17]: check if we can use our logger in here

File: socketio/socketio.go [hunk 18]
```diff
@@ -138,8 +147,28 @@ func ginMiddleware() gin.HandlerFunc {
 	}
 }
 
+func getLogs(hours int) ([]models.Log, error){
```
Comment (idanasulinmemphis) [inline socketio/socketio.go:150]: this function has to sit inside logs handler

File: socketio/socketio.go [hunk 19]
```diff
@@ -176,8 +205,22 @@ func InitializeSocketio(router *gin.Engine) *socketio.Server {
 		return "recv " + stationName
 	})
 
+	server.OnEvent("/", "register_system_logs_data", func (s socketio.Conn) string  {
```
Comment (idanasulinmemphis) [inline socketio/socketio.go:208]: get the message in here and return it

File: main.go (hunk 1 above)
Comment (shay23b) [reply main.go:18]: When I save the file it moves up automaticly

File: socketio/socketio.go [hunk 20]
```diff
@@ -14,24 +14,30 @@
...
+	"go.mongodb.org/mongo-driver/mongo"
 )
 
+var server = socketio.NewServer(nil)
+
 var producersHandler = handlers.ProducersHandler{}
 var consumersHandler = handlers.ConsumersHandler{}
 var auditLogsHandler = handlers.AuditLogsHandler{}
 var stationsHandler = handlers.StationsHandler{}
+var systemLogsCollection *mongo.Collection = db.GetCollection("system_logs")
```
Comment (idanasulinmemphis) [inline socketio/socketio.go:40]: this is unused

File: socketio/socketio.go [hunk 21]
```diff
@@ -14,24 +14,30 @@
...
+
+	"errors"
 	"net/http"
 	"strings"
 	"time"
 
 	"github.com/gin-contrib/cors"
 	"github.com/gin-gonic/gin"
 	socketio "github.com/googollee/go-socket.io"
+	"go.mongodb.org/mongo-driver/mongo"
```
Comment (idanasulinmemphis) [inline socketio/socketio.go:31]: can be deleted

File: socketio/socketio.go [hunk 22]
```diff
@@ -14,24 +14,30 @@
 package socketio
 
 import (
-	"errors"
+	"memphis-control-plane/db"
```
Comment (idanasulinmemphis) [inline socketio/socketio.go:17]: can be deleted

File: broker/broker.go [hunk 23]
```diff
@@ -278,6 +280,40 @@ func ValidateUserCreds(token string) error {
...
+		NoAck: false,
+		Duplicates: dedupWindow,
+	}, nats.MaxWait(10*time.Second))
+	if err != nil && !strings.Contains(err.Error(), "stream name already in use"){
+		return getErrorWithoutNats(err)
+	}
+	return nil
+}
+
+func PublishMessageToStream(subject string, msg []byte) error {
```
Comment (idanasulinmemphis) [inline broker/broker.go:301]: messages are sent to a subject not to a stream

**Review by idanasulinmemphis — DISMISSED** [review]

**Review by idanasulinmemphis — APPROVED** [review]

**Pull request merged**